prow/plugins/config.go

```diff
@@ -730,39 +750,57 @@ func (r RequireMatchingLabel) Describe() string {
 	return str.String()
 }
 
-// ApproveFor finds the Approve for a repo, if one exists.
-// Approval configuration can be listed for a repository
-// or an organization.
-func (c *Configuration) ApproveFor(org, repo string) *Approve {
-	fullName := fmt.Sprintf("%s/%s", org, repo)
-
-	a := func() *Approve {
-		// First search for repo config
-		for _, approve := range c.Approve {
-			if !sets.NewString(approve.Repos...).Has(fullName) {
-				continue
+func oldToNewApprove(old DeprecatedApprove) *Approve {
+	a := Approve{
+		IgnoreReviewState:   old.IgnoreReviewState,
+		IssueRequired:       old.IssueRequired,
+		LgtmActsAsApprove:   old.LgtmActsAsApprove,
+		RequireSelfApproval: old.RequireSelfApproval,
+		CommandHelpLink:     old.CommandHelpLink,
+		PrProcessLink:       old.PrProcessLink,
+	}
+	return &a
+}
+
+func oldToNewApproveConfig(old []DeprecatedApprove) ApproveConfigTree {
+	a := ApproveConfigTree{}
+	a.Orgs = make(map[string]ApproveOrg)
+	for _, entry := range old {
+		for _, repo := range entry.Repos {
+			s := strings.Split(repo, "/")
+			ao := a.Orgs[s[0]]
+			switch len(s) {
+			case 1:
+				ao.Approve = *oldToNewApprove(entry)
+			case 2:
+				if ao.Repos == nil {
+					ao.Repos = make(map[string]ApproveRepo)
+				}
+				ar := ao.Repos[s[1]]
+				ar.Approve = *oldToNewApprove(entry)
+				ao.Repos[s[1]] = ar
 			}
-			return &approve
+			a.Orgs[s[0]] = ao
 		}
+	}
+	return a
+}
 
-		// If you don't find anything, loop again looking for an org config
-		for _, approve := range c.Approve {
-			if !sets.NewString(approve.Repos...).Has(org) {
-				continue
-			}
-			return &approve
-		}
+type approveWithoutUnmarshaler ApproveConfigTree
 
-		// Return an empty config, and use plugin defaults
-		return &Approve{}
-	}()
-	if a.CommandHelpLink == "" {
-		a.CommandHelpLink = "https://go.k8s.io/bot-commands"
-	}
-	if a.PrProcessLink == "" {
-		a.PrProcessLink = "https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process"
+var warnTriggerDeprecatedApprove time.Time
+
+func (a *ApproveConfigTree) UnmarshalJSON(d []byte) error {
```

Comment:
@alvaroaleman I have finally taken some time to implement the transparent unmarshalling...